src/transforms/groupby.js

```diff
@@ -158,20 +152,52 @@ function transformOne(trace, state) {
         styleLookup[styles[i].target] = styles[i].value;
     }
 
+    var newDataByGroup = {};
+
     for(i = 0; i < groupNames.length; i++) {
-        var groupName = groupNames[i];
+        groupName = groupNames[i];
+
+        // Start with a deep extend that just copies array references.
+        newTrace = newData[i] = newDataByGroup[groupName] = Lib.extendDeepNoArrays({}, trace);
+        newTrace.name = groupName;
 
-        var newTrace = newData[i] = Lib.extendDeepNoArrays({}, trace);
+        // In order for groups to apply correctly to other transform data (e.g.
+        // a filter transform), we have to break the connection and clone the
+        // transforms so that each group writes grouped values into a different
+        // destination. This function does not break the array reference
+        // connection between the split transforms it creates. That's handled in
+        // initialize, which creates a new empty array for each arrayAttr.
+        transforms = newTrace.transforms;
+        newTrace.transforms = [];
+        for(j = 0; j < transforms.length; j++) {
+            newTrace.transforms[j] = Lib.extendDeepNoArrays({}, transforms[j]);
+        }
+
+        // Initialize empty arrays for the arrayAttrs, to be split in the next step
+        for(j = 0; j < arrayAttrs.length; j++) {
+            Lib.nestedProperty(newTrace, arrayAttrs[j]).set([]);
```

Comment:
Yeah, that was bugging me, though I know the effect is small in the grand scheme of things. You know exactly what you'll need, so I just indexed them directly and then looked them up by group index. It's still lots of nestedProperty instantiation, but unless we get fancy (read: fragile), I think this is somewhat minimal for the task at hand.